File: bluetooth/src/main/java/de/falkorichter/android/bluetooth/commandQueue/GattCommand.java
package de.falkorichter.android.bluetooth.commandQueue;

import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;

public class GattCommand {

    public enum CommandType {
        TYPE_DESCRIPTOR,
        TYPE_CHARACTERISTIC
    }

    public enum CommandOperation {
        OPERATION_READ,
        CommandOperation, OPERATION_WRITE
    }

    public final CommandType commandType;
    public final CommandOperation commandOperation;
    private BluetoothGattDescriptor bluetoothGattDescriptor;
    private BluetoothGattCharacteristic bluetoothGattCharacteristic;

    public GattCommand(BluetoothGattDescriptor bluetoothGattDescriptor, CommandOperation commandOperation) {
        this.bluetoothGattDescriptor = bluetoothGattDescriptor;
        this.commandOperation = commandOperation;
        this.commandType = CommandType.TYPE_DESCRIPTOR;
    }

    public GattCommand(BluetoothGattCharacteristic bluetoothGattCharacteristic, CommandOperation commandOperation) {
        this.bluetoothGattCharacteristic = bluetoothGattCharacteristic;
        this.commandOperation = commandOperation;
        this.commandType = CommandType.TYPE_CHARACTERISTIC;
    }

    public BluetoothGattDescriptor getBluetoothGattDescriptor() {
        return this.bluetoothGattDescriptor;
    }

    public BluetoothGattCharacteristic getBluetoothGattCharacteristic() {
        return this.bluetoothGattCharacteristic;
    }
}

File: bluetooth/src/main/kotlin/de/falkorichter/android/bluetooth/commandQueue/GattCommand.kt
package de.falkorichter.android.bluetooth.commandQueue

import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor

public fun GattCommand(bluetoothGattDescriptor: BluetoothGattDescriptor, commandOperation: CommandOperation): GattCommand {
    return GattCommand(GattCommand.CommandType.TYPE_DESCRIPTOR, commandOperation, bluetoothGattDescriptor, null)
}

public fun GattCommand(bluetoothGattCharacteristic: BluetoothGattCharacteristic, commandOperation: CommandOperation): GattCommand {
    return GattCommand(GattCommand.CommandType.TYPE_CHARACTERISTIC, commandOperation, null, bluetoothGattCharacteristic)
}

public class GattCommand(
        public val commandType: CommandType, public val commandOperation: CommandOperation, public val bluetoothGattDescriptor: BluetoothGattDescriptor, public val bluetoothGattCharacteristic: BluetoothGattCharacteristic) {

    public enum class CommandType {
        TYPE_DESCRIPTOR
        TYPE_CHARACTERISTIC
    }

    public enum class CommandOperation {
        OPERATION_READ
        CommandOperation
        OPERATION_WRITE
    }
}
